src/pages/DataLibrary/components/UserLibrary/index.tsx

```diff
@@ -597,46 +612,79 @@ const UploadComponent = ({
   function getTimestamp() {
     const pad = (n: any, s = 2) => `${new Array(s).fill(0)}${n}`.slice(-s);
     const d = new Date();
-
     return `${pad(d.getFullYear(), 4)}-${pad(d.getMonth() + 1)}-${pad(
       d.getDate()
     )}-${pad(d.getHours())}:${pad(d.getMinutes())}:${pad(d.getSeconds())}`;
   }
 
+  const uploadFile = async (
+    file: File,
+    url: string,
+    onUploadProgress: (progressEvent: ProgressEvent) => void
+  ) => {
+    const formData = new FormData();
+    formData.append(
+      "upload_path",
+      `${username}/uploads/${directoryName}/${file.name}`
+    );
+    formData.append("fname", file, file.name);
+
+    const config = {
+      headers: { Authorization: "Token " + token },
+      onUploadProgress,
+    };
+
+    const response = await axios.post(url, formData, config);
+    return response;
+  };
+
   React.useEffect(() => {
     const d = getTimestamp();
     setDirectoryName(`${d}`);
   }, [uploadFileModal]);
 
   const handleUpload = async () => {
     const client = ChrisAPIClient.getClient();
-    const setDirectory = !directoryName ? getTimestamp() : directoryName;
-    handleAddFolder(setDirectory);
-    const path = `${username}/uploads/${setDirectory}`;
+    await client.setUrls();
+    const url = client.uploadedFilesUrl;
+
+    const fileUploads: FileUpload[] = localFiles.map((file) => {
+      const onUploadProgress = (progressEvent: ProgressEvent) => {
+        const percentCompleted = `${Math.round(
+          (progressEvent.loaded * 100) / progressEvent.total
+        )}%`;
+        setCurrentFile((prevProgresses) => ({
+          ...prevProgresses,
+          [file.name]: percentCompleted,
+        }));
+      };
+      const promise = uploadFile(file, url, onUploadProgress);
+
+      return {
+        file,
+        promise,
+      };
+    });
 
-    try {
-      for (let i = 0; i < localFiles.length; i++) {
-        const file = localFiles[i];
-
-        await client.uploadFile(
-          {
-            upload_path: `${path}/${file.name}`,
-          },
-          {
-            fname: file,
-          }
-        );
-        setCount(i + 1);
+    const completedUploads: number[] = [];
+
+    for (let i = 0; i < fileUploads.length; i++) {
+      const { promise } = fileUploads[i];
+      try {
+        await promise;
```

Comment:
`await` inside of a loop should be avoided, we could try to do concurrent uploads using `Promise.all`, or use some thread-pool-like helper function